### PR TITLE
fix: allow to parse "-" as a key code

### LINKED ISSUE #12190: Invalid key code when mapping "-"
### Summary

Trying to map a `"-"` key to an action causes parsing failure.

Looking at the code, this is because `'-'` is used as a modifier separator and `"minus"` should be used instead. This restriction is not clear from docs alone, but imho rather than documenting it, it would be better to improve the parser to allow it as there is no ambiguity:

- without a modifier `"-"`: on it's own it is clear we're binding to this char
- with a modifier `"A--": arguably not as readable as `"A-minus"`, but the middle '-' must be a separator and the '-' the key code
- invalid binding `"A-"`: missing key code

### Reproduction Steps

trying to map a "-" key to an action with e.g.:

```toml
[keys.normal.space.w]
"-" = "hsplit"
```

causes a failure

```
Bad config: TOML parse error at line x, column y
    |
x | "-" = "hsplit"
    | ^^^
Invalid key code ''
```

### Helix log

_No response_

### Platform

cross-platform

### Terminal Emulator

any

### Installation Method

source

### Helix Version

helix 24.7 (cf81e153)

### PR DESCRIPTION
**Pull request by tzemanovic**
closes #12190

### CODE REVIEW

**Comment by the-mikedavis** [comment]
Parsing `-` is ok (also see https://github.com/helix-editor/helix/pull/8475 which accepts it for macros) but I'd like to continue rejecting `<Mod>--` since it's not very readable.

**Comment by tzemanovic** [comment]
> Parsing `-` is ok (also see #8475 which accepts it for macros) but I'd like to continue rejecting `<Mod>--` since it's not very readable.

sounds good, thanks! I made the change

File: helix-view/src/input.rs [hunk 1]
```diff
@@ -387,6 +387,21 @@ impl std::str::FromStr for KeyEvent {
                     .then_some(KeyCode::F(function))
                     .ok_or_else(|| anyhow!("Invalid function key '{}'", function))?
             }
+            _ if s.ends_with('-')
+                && tokens.last().map(|s| str::is_empty(s)).unwrap_or_default() =>
+            {
+                if tokens.len() > 1 {
```
Comment (the-mikedavis) [inline helix-view/src/input.rs:393]: I think the tokens stuff is a little hard to follow. We can check for the "is `-` and doesn't have modifiers" case easier with `s == "-"` and then use `tokens.clear()` and return `KeyCode::Char('-')` in that branch.

I like the explicit error for this case recommending using `keys::MINUS`. We can expand on it a little by recommending the full key sequence:

```rust
let mut suggestion = s[..s.len() - 1].to_string();
suggestion.push_str(keys::MINUS);
```

and suggesting that in the error message

File: helix-view/src/input.rs [hunk 2]
```diff
@@ -387,6 +387,21 @@ impl std::str::FromStr for KeyEvent {
                     .then_some(KeyCode::F(function))
                     .ok_or_else(|| anyhow!("Invalid function key '{}'", function))?
             }
+            _ if s.ends_with('-')
+                && tokens.last().map(|s| str::is_empty(s)).unwrap_or_default() =>
```
Comment (the-mikedavis) [inline helix-view/src/input.rs:391]:
```suggestion
                && tokens.last().map_or(false, |s| s.is_empty()) =>
```

File: helix-view/src/input.rs [hunk 3]
```diff
@@ -387,6 +387,21 @@ impl std::str::FromStr for KeyEvent {
...
+            {
+                if tokens.len() > 1 {
+                    return Err(anyhow!(
+                        "Key '-' cannot be used with modifiers, use '{}' instead",
+                        keys::MINUS
+                    ));
```
Comment (the-mikedavis) [inline helix-view/src/input.rs:397]: I like the explicit error for this case recommending using `keys::MINUS`. We can expand on it a little by recommending the full key sequence:

```rust
let mut suggestion = s[..s.len() - 1].to_string();
suggestion.push_str(keys::MINUS);
```

and suggesting that in the error message

File: helix-view/src/input.rs (hunk 1 above)
Comment (tzemanovic) [reply helix-view/src/input.rs]: good point - this can be a lot simpler now

File: helix-view/src/input.rs (hunk 2 above)
Comment (tzemanovic) [reply helix-view/src/input.rs]: this is now gone from applying the previous suggestion

**Review by the-mikedavis — APPROVED** [review]

**Comment by eugenesvk** [comment]
> rejecting `<Mod>--` since it's not very readable

why can't the people who can't read it not use it? Since currently the shift from a symbol to a word is less readable as it breaks the visual similarity and you have an extra translation step

```
  '-'    	= 'trim_selections'             	#
'A--'    	= 'merge_selections'            	#
'A-minus'	= 'merge_selections'            	#
  '_'    	= 'merge_consecutive_selections'	# poor readability, but
'S--'    	= 'merge_consecutive_selections'	# a separate issue
```
